File: system/modules/watchdog/notifications/pushover_client.py
"""
Pushover API client for Watchdog notifications.

Matches Robot's PushoverClient behavior: same endpoint, priority mapping.
Uses urllib for no extra dependencies.
"""
import logging
import urllib.error
import urllib.parse
import urllib.request
from typing import Optional, Tuple

logger = logging.getLogger(__name__)

PUSHOVER_ENDPOINT = "https://api.pushover.net/1/messages.json"
DEFAULT_TIMEOUT = 12


def pushover_send(
    user_key: str,
    app_token: str,
    title: str,
    message: str,
    priority: int = 0,
    timeout: float = DEFAULT_TIMEOUT,
) -> Tuple[bool, Optional[int], Optional[str]]:
    """
    Send a push notification via Pushover API.

    Args:
        user_key: Pushover user key
        app_token: Pushover application token
        title: Notification title
        message: Notification message body
        priority: 0=normal, 1=high, 2=emergency
        timeout: Request timeout in seconds

    Returns:
        Tuple of (success, http_status_code, error_message)
    """
    if not user_key or not app_token:
        return False, None, "User key or app token is empty"

    data = {
        "token": app_token,
        "user": user_key,
        "title": title[:250] if title else "",
        "message": message[:1024] if message else "",
        "priority": priority,
    }
    if priority == 2:
        data["expire"] = 3600
        data["retry"] = 60

    body = urllib.parse.urlencode(data).encode("utf-8")
    req = urllib.request.Request(
        PUSHOVER_ENDPOINT,
        data=body,
        method="POST",
        headers={"Content-Type": "application/x-www-form-urlencoded"},
    )

    try:
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            status = resp.getcode()
            if 200 <= status < 300:
                return True, status, None
            return False, status, f"HTTP {status}"
    except urllib.error.HTTPError as e:
        return False, e.code, str(e)
    except urllib.error.URLError as e:
        return False, None, str(e.reason) if e.reason else str(e)
    except TimeoutError as e:
        return False, None, str(e)
    except Exception as e:
        logger.exception("Pushover send failed")
        return False, None, str(e)

File: system/RobotCore_For_NinjaTrader/Notifications/PushoverClient.cs
using System;
using System.Net.Http;
using System.Text;
using System.Threading.Tasks;

namespace QTSW2.Robot.Core.Notifications;

/// <summary>
/// Pushover API client for sending push notifications.
/// Non-blocking: returns Task, never awaited on hot paths.
/// </summary>
public static class PushoverClient
{
    public const string PUSHOVER_ENDPOINT = "https://api.pushover.net/1/messages.json";
    private static readonly HttpClient _httpClient = new HttpClient
    {
        Timeout = TimeSpan.FromSeconds(12)  // Slightly longer than NotificationService guard timeout (10s) to ensure guard timeout triggers first
    };

    /// <summary>
    /// Result of a Pushover API call with full diagnostic details.
    /// </summary>
    public class SendResult
    {
        public bool Success { get; set; }
        public int? HttpStatusCode { get; set; }
        public string? ResponseBody { get; set; }
        public Exception? Exception { get; set; }
    }

    /// <summary>
    /// Send a push notification via Pushover API.
    /// Non-blocking: returns Task that should be fire-and-forget.
    /// </summary>
    /// <param name="userKey">Pushover user key</param>
    /// <param name="appToken">Pushover application token</param>
    /// <param name="title">Notification title</param>
    /// <param name="message">Notification message</param>
    /// <param name="priority">Priority (0 = normal, 1 = high, 2 = emergency)</param>
    /// <returns>SendResult with success status and full diagnostic details</returns>
    public static async Task<SendResult> SendAsync(string userKey, string appToken, string title, string message, int priority = 0)
    {
        var result = new SendResult();
        
        if (string.IsNullOrWhiteSpace(userKey) || string.IsNullOrWhiteSpace(appToken))
        {
            result.Success = false;
            result.Exception = new ArgumentException("User key or app token is null or empty");
            return result;
        }

        try
        {
            var formData = new StringBuilder();
            formData.Append($"token={Uri.EscapeDataString(appToken)}");
            formData.Append($"&user={Uri.EscapeDataString(userKey)}");
            formData.Append($"&title={Uri.EscapeDataString(title)}");
            formData.Append($"&message={Uri.EscapeDataString(message)}");
            formData.Append($"&priority={priority}");
            
            // Priority 2 (emergency) requires both expire and retry parameters
            // expire: maximum time in seconds that the notification will be retried (max 10800 = 3 hours)
            // retry: how often in seconds the notification will be retried (min 30 seconds)
            // Set expire to 3600 seconds (1 hour) and retry to 60 seconds (retry every minute)
            if (priority == 2)
            {
                formData.Append($"&expire=3600");
                formData.Append($"&retry=60");
            }

            var content = new StringContent(formData.ToString(), Encoding.UTF8, "application/x-www-form-urlencoded");

            var response = await _httpClient.PostAsync(PUSHOVER_ENDPOINT, content);
            
            result.HttpStatusCode = (int)response.StatusCode;
            result.ResponseBody = await response.Content.ReadAsStringAsync();
            
            if (!response.IsSuccessStatusCode)
            {
                result.Success = false;
                result.Exception = new HttpRequestException($"Pushover API returned non-success status code: {response.StatusCode}");
            }
            else
            {
                result.Success = true;
            }
        }
        catch (HttpRequestException ex)
        {
            result.Success = false;
            result.Exception = ex;
        }
        catch (TaskCanceledException ex)
        {
            result.Success = false;
            result.Exception = ex;
        }
        catch (Exception ex)
        {
            result.Success = false;
            result.Exception = ex;
        }
        
        return result;
    }
}
